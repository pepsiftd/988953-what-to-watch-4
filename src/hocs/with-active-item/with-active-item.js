import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

const withActiveItem = (Component) => {
  class WithActiveItem extends PureComponent {
    constructor(props) {
      super(props);
      this.state = {
        currentItemId: props.initialItemId ? props.initialItemId : null
      };
    }

    render() {
      const currentItemId = this.state.currentItemId;

      return (
        <Component
          {...this.props}
          renderItem={(ItemComponent, itemId, itemProps) => {
            return (
              <ItemComponent
                {...itemProps}
                isActive={itemId === currentItemId}
              />
            );
          }}
          setActiveItem={(id) => {
            this.setState({
              currentItemId: id
            });
          }}
          clearActiveItem={() => {
            this.setState({
              currentItemId: null
            });
          }}
        />
      );
    }
  }

  WithActiveItem.propTypes = {
    initialItemId: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  };

  return WithActiveItem;
};

export {withActiveItem};
