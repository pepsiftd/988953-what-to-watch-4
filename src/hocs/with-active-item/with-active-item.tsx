import React, {PureComponent} from 'react';
import {Subtract} from 'utility-types';

interface State {
  currentItemId: number | string;
}

interface InjectingProps {
  activeItemId: number | string;
  renderItem: (ItemComponent: React.ComponentClass, itemId: number | string, itemProps: Record<string, unknown>) => React.ReactNode;
  setActiveItem: (id: string | number) => void;
  clearActiveItem: () => void;
}

const withActiveItem: (Component: React.ComponentClass) => React.ReactNode = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithActiveItem extends PureComponent<T, State> {
    constructor(props) {
      super(props);
      this.state = {
        currentItemId: props.initialItemId ? props.initialItemId : null
      };
    }

    render() {
      const currentItemId = this.state.currentItemId;

      return (
        <Component
          {...this.props}
          activeItemId={currentItemId}
          renderItem={(ItemComponent, itemId, itemProps) => {
            return (
              <ItemComponent
                {...itemProps}
                isActive={itemId === currentItemId}
              />
            );
          }}
          setActiveItem={(id) => {
            this.setState({
              currentItemId: id
            });
          }}
          clearActiveItem={() => {
            this.setState({
              currentItemId: null
            });
          }}
        />
      );
    }
  }

  return WithActiveItem;
};

export {withActiveItem};
