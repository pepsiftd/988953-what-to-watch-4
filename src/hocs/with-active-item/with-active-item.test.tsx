import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {withActiveItem} from './with-active-item';

interface ItemProps {
  isActive: boolean;
}

const MockComponentItem: React.FunctionComponent<ItemProps> = ({isActive}) => {
  return (
    <li className={`list-item ${isActive ? `list-item--active` : ``}`} />
  );
};

interface Props {
  renderItem: (Component: React.FunctionComponent | React.ComponentClass, id: number | string, props: Record<string, unknown>) => React.ReactNode;
}

const MockComponent: React.FunctionComponent<Props> = ({renderItem}) => {
  return (
    <ul>
      {renderItem(MockComponentItem, 1, {key: 1})}
      {renderItem(MockComponentItem, 2, {key: 2})}
    </ul>
  );
};

it(`withActiveItem renders Active-Inactive items correctly`, () => {
  const MockComponentWrapped = withActiveItem(MockComponent);

  const tree = renderer.create(
      <MockComponentWrapped
        initialItemId={1}
      />
  );

  expect(tree).toMatchSnapshot();
});
