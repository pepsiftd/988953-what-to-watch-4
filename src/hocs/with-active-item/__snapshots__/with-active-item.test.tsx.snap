// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withActiveItem renders Active-Inactive items correctly 1`] = `
<ul>
  <li
    className="list-item list-item--active"
  />
  <li
    className="list-item "
  />
</ul>
`;
